// @bhplugin/ng-datatable/esm2020/lib/ng-datatable.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { NgDataTableComponent } from './ng-datatable';
import { ColumnFilterComponent } from './column-filter';
import { ColumnHeaderComponent } from './column-header';
import { IconCheckComponent } from './icon-check';
import { IconDashComponent } from './icon-dash';
import { IconFilterComponent } from './icon-filter';
import { IconLoaderComponent } from './icon-loader';
// directive
import { SlotDirective } from './slot.directive';
import * as i0 from "@angular/core";
export class DataTableModule {
}
DataTableModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: DataTableModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
DataTableModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.9", ngImport: i0, type: DataTableModule, declarations: [NgDataTableComponent,
        ColumnFilterComponent,
        ColumnHeaderComponent,
        IconCheckComponent,
        IconDashComponent,
        IconFilterComponent,
        IconLoaderComponent,
        SlotDirective], imports: [CommonModule, FormsModule], exports: [NgDataTableComponent, ColumnFilterComponent, ColumnHeaderComponent, IconCheckComponent, IconDashComponent, IconFilterComponent, IconLoaderComponent, SlotDirective] });
DataTableModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: DataTableModule, imports: [CommonModule, FormsModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.9", ngImport: i0, type: DataTableModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule, FormsModule],
                    declarations: [
                        NgDataTableComponent,
                        ColumnFilterComponent,
                        ColumnHeaderComponent,
                        IconCheckComponent,
                        IconDashComponent,
                        IconFilterComponent,
                        IconLoaderComponent,
                        SlotDirective,
                    ],
                    exports: [NgDataTableComponent, ColumnFilterComponent, ColumnHeaderComponent, IconCheckComponent, IconDashComponent, IconFilterComponent, IconLoaderComponent, SlotDirective],
                }]
        }] });
